consoles/VNC.pm

```diff
@@ -921,6 +931,63 @@ sub _receive_zrle_encoding ($self, $x, $y, $w, $h) {
     return $res;
 }
 
+sub _receive_tight_encoding ($self, $x, $y, $w, $h) {
+    my $socket = $self->socket;
+
+    my $pi = $self->_pixinfo;
+
+    my $stime = time;
+    $socket->read(my $data, 1)
+      or OpenQA::Exception::VNCProtocolError->throw(error => 'short read for compression control');
+    my ($compression_control) = unpack('C', $data);
+    # FillCompression
+    if (($compression_control & 0xF0) eq 0x80) {
+        my $data;
+        # special case for TPIXEL, otherwise identical to PIXEL
+        if ($self->_true_colour and $self->_bpp eq 32 and $self->depth eq 24) {
+            $socket->read($data, 3)
+              or OpenQA::Exception::VNCProtocolError->throw(error => 'short read for compression control');
+            $data = pack('CCCx', unpack('CCC', $data));
+        } else {
+            $socket->read($data, $self->_bpp / 8)
+              or OpenQA::Exception::VNCProtocolError->throw(error => 'short read for compression control');
+        }
+        my $image = $self->_framebuffer;
+        $image->fill_pixel($data, $self->vncinfo, $x, $y, $w, $h);
+        return;
+    }
+    # Only Fill (above) and JPEG for now; "Basic" unsupported
+    die "Unsupported compression $compression_control" if ($compression_control & 0xF0) != 0x90;
+
+    $socket->read($data, 1)
+      or OpenQA::Exception::VNCProtocolError->throw(error => 'short read for data len');
+    my ($data_len) = unpack('C', $data);
+    if ($data_len & 0x80) {
+        $socket->read($data, 1)
+          or OpenQA::Exception::VNCProtocolError->throw(error => 'short read for data len');
+        my ($data_len2) = unpack('C', $data);
+        $data_len = ($data_len & 0x7f) | ($data_len2 & 0x7f) << 7;
+        if ($data_len2 & 0x80) {
+            $socket->read($data, 1)
+              or OpenQA::Exception::VNCProtocolError->throw(error => 'short read for data len');
+            my ($data_len2) = unpack('C', $data);
+            $data_len |= $data_len2 << 14;
+        }
+    }
+
+    $data = "";
+    my $read_len = 0;
+    while ($read_len < $data_len) {
+        my $len = $socket->read(my $data_tmp, $data_len - $read_len);
```

Comment:
@Martchus I changed this from `read($socket, $data, $data_len - $read_len, $read_len)`, but I'm not entirely happy about the change (especially the `$data .= $data_tmp` below). But I couldn't figure out how to mock the old variant for tests

Reply:
I suppose you could always just wrap the call to the built-in `read(…)` function in another function and mock that.